src/celeritas/optical/model/RayleighModel.cc

```diff
@@ -0,0 +1,116 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/model/RayleighModel.cc
+//---------------------------------------------------------------------------//
+#include "RayleighModel.hh"
+
+#include <iostream>
+
+#include "corecel/Assert.hh"
+#include "corecel/io/Logger.hh"
+#include "celeritas/Constants.hh"
+#include "celeritas/io/ImportOpticalMaterial.hh"
+#include "celeritas/optical/MfpBuilder.hh"
+
+namespace celeritas
+{
+namespace optical
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the model from imported data.
+ */
+RayleighModel::RayleighModel(ActionId id, SPConstImported imported, Input input)
+    : Model(id, "optical-rayleigh", "interact by optical Rayleigh")
+    , imported_(ImportModelClass::rayleigh, imported)
+    , properties_(std::move(input.properties))
+    , rayleigh_(std::move(input.rayleigh))
+{
+    CELER_EXPECT(rayleigh_.size() == properties_.size());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Build the mean free paths for the model.
+ */
+void RayleighModel::build_mfps(OpticalMaterialId mat, MfpBuilder& build) const
+{
+    using namespace celeritas::constants;
+
+    if (auto const& mfp = imported_.mfp(mat))
+    {
+        build(mfp);
+    }
+    else if (auto const& rayl = rayleigh_[mat.get()])
+    {
+        ImportPhysicsVector const& r_index
+            = properties_[mat.get()].refractive_index;
+
+        constexpr real_type hbarc = hbar_planck * c_light;
+
+        // Einstein-Smoluchowski formula
+        real_type scale_factor = rayl.scale_factor;
+        real_type beta_t = rayl.compressibility;
+        real_type temperature = rayl.temperature;
+
+        real_type c1 = scale_factor * beta_t * temperature * k_boltzmann
+                       / (6 * pi);
+
+        ImportPhysicsVector calculated_mfp{
+            ImportPhysicsVectorType::free,
+            r_index.x,
+            std::vector<double>(r_index.y.size(), 0)};
+        for (auto i : range(r_index.x.size()))
+        {
+            real_type energy = r_index.x[i];
+            real_type n = r_index.y[i];
+
+            CELER_ASSERT(n > 1);
+
+            real_type c2 = ipow<4>(energy / hbarc);
+            real_type c3 = ipow<2>((ipow<2>(n) - 1) * (ipow<2>(n) + 2) / 3);
+
+            CELER_ASSERT(c2 > 0);
+            CELER_ASSERT(c3 > 0);
+
+            calculated_mfp.y[i] = 1 / (c1 * c2 * c3);
```

Comment:
This should be done in a separate `RayleighMfpCalculator` class, using `optical::MaterialView::make_refractive_index_calculator()`. I think the class declaration should probably be:
```c++
class RayleighMfpCalculator
{
  public:
    //!@{
    //! \name Type aliases
    using Energy = units::MevEnergy;
    //!@}

  public:
    inline CELER_FUNCTION RayleighMfpCalculator(MaterialView const& mat,
                                                real_type scale_factor,
                                                real_type compressibility);

    inline CELER_FUNCTION real_type operator()(Energy) const;

  private:
    GenericCalculator calc_rindex_;
    real_type c1_;
};
```

There shouldn't be a hard requirement to calculate the rayleigh cross sections on the same grid as the rindex (though that is likely the sensible way to start).